File: src/donorNeighbourhoodArea.js
import { normalizeLocalityKey } from "./localityKey.js";

const DEFAULT_RADIUS_M = 5000;
/** Upper bound only — prevents oversized ST_DWithin scans; no high minimum floor. */
const MAX_RADIUS_M = 50_000;

/**
 * @param {string | number | undefined} raw
 * @returns {number} metres
 */
export function parseDonorNeighbourhoodRadiusM(raw) {
  if (raw == null || raw === "") {
    return DEFAULT_RADIUS_M;
  }
  const parsed = Number(String(raw).trim());
  if (!Number.isFinite(parsed) || parsed <= 0) {
    return DEFAULT_RADIUS_M;
  }
  return Math.min(MAX_RADIUS_M, Math.round(parsed));
}

/** From `INITIATOR_NEIGHBOURHOOD_RADIUS_M` (default 5000 m). */
export function getDonorNeighbourhoodRadiusM() {
  return parseDonorNeighbourhoodRadiusM(
    process.env.INITIATOR_NEIGHBOURHOOD_RADIUS_M
  );
}

/**
 * @param {string | null | undefined} value
 * @param {"lat" | "lng"} kind
 * @returns {number | null}
 */
export function parseGeoCoord(value, kind) {
  if (value == null || value === "") {
    return null;
  }
  const n = Number(String(value).trim());
  if (!Number.isFinite(n)) {
    return null;
  }
  if (kind === "lat") {
    return n >= -90 && n <= 90 ? n : null;
  }
  return n >= -180 && n <= 180 ? n : null;
}

/**
 * @param {URLSearchParams} params
 */
export function parseNeighbourhoodQuery(params) {
  const nearLat = parseGeoCoord(params.get("near_lat"), "lat");
  const nearLng = parseGeoCoord(params.get("near_lng"), "lng");
  const localityKey = normalizeLocalityKey(params.get("locality_key") ?? "");
  let radiusM = getDonorNeighbourhoodRadiusM();
  const radiusRaw = params.get("radius_m");
  if (radiusRaw != null && radiusRaw !== "") {
    const parsed = Number(String(radiusRaw).trim());
    if (Number.isFinite(parsed) && parsed > 0) {
      radiusM = Math.min(Math.round(parsed), getDonorNeighbourhoodRadiusM());
    }
  }
  return {
    nearLat,
    nearLng,
    localityKey: localityKey || null,
    radiusM
  };
}

File: src/main/java/org/sharingbridge/integration/geo/DonorNeighbourhoodArea.java
package org.sharingbridge.integration.geo;

public final class DonorNeighbourhoodArea {

    public static final int DEFAULT_RADIUS_M = 5000;
    public static final int MAX_RADIUS_M = 50_000;

    private DonorNeighbourhoodArea() {}

    public static int parseDonorNeighbourhoodRadiusM(Object raw) {
        if (raw == null || String.valueOf(raw).trim().isEmpty()) {
            return DEFAULT_RADIUS_M;
        }
        try {
            double parsed = Double.parseDouble(String.valueOf(raw).trim());
            if (!Double.isFinite(parsed) || parsed <= 0) {
                return DEFAULT_RADIUS_M;
            }
            return Math.min(MAX_RADIUS_M, (int) Math.round(parsed));
        } catch (NumberFormatException ex) {
            return DEFAULT_RADIUS_M;
        }
    }

    public static int getDonorNeighbourhoodRadiusM() {
        return parseDonorNeighbourhoodRadiusM(System.getenv("INITIATOR_NEIGHBOURHOOD_RADIUS_M"));
    }

    public static Double parseGeoCoord(Object value, String kind) {
        if (value == null || String.valueOf(value).trim().isEmpty()) {
            return null;
        }
        try {
            double n = Double.parseDouble(String.valueOf(value).trim());
            if (!Double.isFinite(n)) {
                return null;
            }
            if ("lat".equals(kind)) {
                return n >= -90 && n <= 90 ? n : null;
            }
            return n >= -180 && n <= 180 ? n : null;
        } catch (NumberFormatException ex) {
            return null;
        }
    }

    public record NeighbourhoodQuery(
            Double nearLat, Double nearLng, String localityKey, int radiusM) {}

    public static NeighbourhoodQuery parseNeighbourhoodQuery(
            String nearLatRaw, String nearLngRaw, String localityKeyRaw, String radiusRaw) {
        Double nearLat = parseGeoCoord(nearLatRaw, "lat");
        Double nearLng = parseGeoCoord(nearLngRaw, "lng");
        String localityKey = LocalityKey.normalizeLocalityKey(localityKeyRaw == null ? "" : localityKeyRaw);
        int radiusM = getDonorNeighbourhoodRadiusM();
        if (radiusRaw != null && !radiusRaw.isBlank()) {
            try {
                double parsed = Double.parseDouble(radiusRaw.trim());
                if (Double.isFinite(parsed) && parsed > 0) {
                    radiusM = Math.min((int) Math.round(parsed), getDonorNeighbourhoodRadiusM());
                }
            } catch (NumberFormatException ignored) {
                // keep default
            }
        }
        return new NeighbourhoodQuery(
                nearLat, nearLng, localityKey.isEmpty() ? null : localityKey, radiusM);
    }
}
